File: Unity_SDK/android_build/unity_chat_sdk/src/main/java/com/hyphenate/unity_chat_sdk/EMContactManagerWrapper.java
package com.hyphenate.unity_chat_sdk;

import android.util.Log;

import com.hyphenate.chat.EMClient;
import com.hyphenate.exceptions.HyphenateException;
import com.hyphenate.unity_chat_sdk.helper.EMTransformHelper;
import com.hyphenate.unity_chat_sdk.listeners.EMUnityContactListener;

import org.json.JSONArray;
import org.json.JSONException;

import java.util.List;

public class EMContactManagerWrapper extends  EMWrapper {

    static public EMContactManagerWrapper wrapper() {
        return new EMContactManagerWrapper();
    }

    public EMContactManagerWrapper() {
        EMClient.getInstance().contactManager().setContactListener(new EMUnityContactListener());
    }

    private void addContact(String username, String reason, String callbackId) throws JSONException {
        asyncRunnable(() -> {
            if (username == null || username.length() == 0) {
                onError(callbackId, new HyphenateException(501, "username invalid"));
                return;
            }
            try {
                EMClient.getInstance().contactManager().addContact(username, reason);
                onSuccess(null, callbackId,null);
            } catch (HyphenateException e) {
                onError(callbackId, e);
            }
        });
    }

    private void deleteContact(String username, boolean keepConversation, String callbackId) throws JSONException {
        asyncRunnable(() -> {
            if (username == null || username.length() == 0) {
                onError(callbackId, new HyphenateException(501, "username invalid"));
                return;
            }
            try {
                EMClient.getInstance().contactManager().deleteContact(username, keepConversation);
                onSuccess(null, callbackId, null);
            } catch (HyphenateException e) {
                onError(callbackId, e);
            }
        });
    }

    private void getAllContactsFromServer(String callbackId) throws JSONException {
        Log.d("chat_sdk", "getAllContactsFromServer");
        asyncRunnable(() -> {
            try {
                List<String> contacts = EMClient.getInstance().contactManager().getAllContactsFromServer();
                Log.d("chat_sdk", contacts.toString());
                onSuccess("List<String>", callbackId, EMTransformHelper.jsonArrayFromStringList(contacts).toString());
            } catch (HyphenateException e) {
                onError(callbackId, e);
            }
        });
    }

    private void getAllContactsFromDB(String callbackId) throws JSONException {
        Log.d("chat_sdk", "getAllContactsFromDB");
        asyncRunnable(() -> {
            try {
                List<String> contacts = EMClient.getInstance().contactManager().getContactsFromLocal();
                Log.d("chat_sdk", contacts.toString());
                onSuccess("List<String>", callbackId, EMTransformHelper.jsonArrayFromStringList(contacts).toString());
            } catch (HyphenateException e) {
                onError(callbackId, e);
            }
        });
    }

    private void addUserToBlockList(String username, String callbackId) throws JSONException {
        Log.d("chat_sdk", "addUserToBlockList");
        asyncRunnable(() -> {
            if (username == null || username.length() == 0) {
                onError(callbackId, new HyphenateException(501, "username invalid"));
                return;
            }
            try {
                EMClient.getInstance().contactManager().addUserToBlackList(username, false);
                onSuccess(null,callbackId,null);
            } catch (HyphenateException e) {
                onError(callbackId, e);
            }
        });
    }

    private void removeUserFromBlockList(String username, String callbackId) throws JSONException {
        asyncRunnable(() -> {
            if (username == null || username.length() == 0) {
                onError(callbackId, new HyphenateException(501, "username invalid"));
                return;
            }
            try {
                EMClient.getInstance().contactManager().removeUserFromBlackList(username);
                onSuccess(null, callbackId,null);
            } catch (HyphenateException e) {
                onError(callbackId, e);
            }
        });
    }

    private void getBlockListFromServer(String callbackId) throws JSONException {
        asyncRunnable(() -> {
            try {
                List<String> contacts = EMClient.getInstance().contactManager().getBlackListFromServer();
                onSuccess("List<String>", callbackId, EMTransformHelper.jsonArrayFromStringList(contacts).toString());
            } catch (HyphenateException e) {
                onError(callbackId, e);
            }
        });
    }

    private void acceptInvitation(String username, String callbackId) throws JSONException {
        asyncRunnable(() -> {
            if (username == null || username.length() == 0) {
                onError(callbackId, new HyphenateException(501, "username invalid"));
                return;
            }
            try {
                EMClient.getInstance().contactManager().acceptInvitation(username);
                onSuccess(null,callbackId, null);
            } catch (HyphenateException e) {
                onError(callbackId, e);
            }
        });
    }

    private void declineInvitation(String username, String callbackId) throws JSONException {
        asyncRunnable(() -> {
            if (username == null || username.length() == 0) {
                onError(callbackId, new HyphenateException(501, "username invalid"));
                return;
            }
            try {
                EMClient.getInstance().contactManager().declineInvitation(username);
                onSuccess(null, callbackId,null);
            } catch (HyphenateException e) {
                onError(callbackId, e);
            }
        });
    }

    private void getSelfIdsOnOtherPlatform(String callbackId) throws JSONException {
        asyncRunnable(() -> {
            try {
                List<String> platforms = EMClient.getInstance().contactManager().getSelfIdsOnOtherPlatform();
                onSuccess("List<String>", callbackId, EMTransformHelper.jsonArrayFromStringList(platforms).toString());
            } catch (HyphenateException e) {
                onError(callbackId, e);
            }
        });
    }
}

File: ChatSDKDemo/Assets/ChatSDK/Plugins/iOS/HyphenateWrapper/Wrapper/EMContactManagerWrapper.m
//
//  EMContactManagerWrapper.m
//  HyphenateWrapper
//
//  Created by 杜洁鹏 on 2021/6/5.
//

#import "EMContactManagerWrapper.h"
#import <HyphenateChat/HyphenateChat.h>
#import "EMContactListener.h"
#import "Transfrom.h"

@interface EMContactManagerWrapper ()<EMContactManagerDelegate>
{
    EMContactListener *_listener;
}
@end

@implementation EMContactManagerWrapper
- (instancetype)init {
    if (self = [super init]) {
        _listener = [[EMContactListener alloc] init];
        [EMClient.sharedClient.contactManager addDelegate:_listener delegateQueue:nil];
    }
    return self;
}

- (void)acceptInvitation:(NSDictionary *)param callbackId:(NSString *)callbackId {
    NSString *username = param[@"username"];
    if (!username) {
        EMError *aError = [[EMError alloc] initWithDescription:@"username is invalid " code: EMErrorMessageIncludeIllegalContent];
        [self onError:callbackId error:aError];
        return;
    }
    __weak EMContactManagerWrapper * weakSelf = self;
    __block NSString *callId = callbackId;
    [EMClient.sharedClient.contactManager approveFriendRequestFromUser:username
                                                            completion:^(NSString *aUsername, EMError *aError)
    {
        if (!aError) {
            [weakSelf onSuccess:nil callbackId:callId userInfo:nil];
        }else {
            [weakSelf onError:callId error:aError];
        }
    }];
}

- (void)addContact:(NSDictionary *)param callbackId:(NSString *)callbackId {
    NSString *username = param[@"username"];
    if (!username) {
        EMError *aError = [[EMError alloc] initWithDescription:@"username is invalid " code: EMErrorMessageIncludeIllegalContent];
        [self onError:callbackId error:aError];
        return;
    }
    
    NSString *reason = param[@"reason"];
    __block NSString *callId = callbackId;
    __weak EMContactManagerWrapper * weakSelf = self;
    [EMClient.sharedClient.contactManager addContact:username
                                             message:reason
                                          completion:^(NSString *aUsername, EMError *aError)
    {
        if (!aError) {
            [weakSelf onSuccess:nil callbackId:callId userInfo:nil];
        }else {
            [weakSelf onError:callbackId error:aError];
        }
    }];
}

- (void)addUserToBlockList:(NSDictionary *)param callbackId:(NSString *)callbackId {
    NSString *username = param[@"username"];
    if (!username) {
        EMError *aError = [[EMError alloc] initWithDescription:@"username is invalid " code: EMErrorMessageIncludeIllegalContent];
        [self onError:callbackId error:aError];
        return;
    }
    __block NSString *callId = callbackId;
    __weak EMContactManagerWrapper * weakSelf = self;
    [EMClient.sharedClient.contactManager addUserToBlackList:username
                                                  completion:^(NSString *aUsername, EMError *aError)
    {
        if (!aError) {
            [weakSelf onSuccess:nil callbackId:callId userInfo:nil];
        }else {
            [weakSelf onError:callId error:aError];
        }
    }];
}

- (void)declineInvitation:(NSDictionary *)param callbackId:(NSString *)callbackId {
    NSString *username = param[@"username"];
    if (!username) {
        EMError *aError = [[EMError alloc] initWithDescription:@"username is invalid " code: EMErrorMessageIncludeIllegalContent];
        [self onError:callbackId error:aError];
        return;
    }
    __weak EMContactManagerWrapper * weakSelf = self;
    __block NSString *callId = callbackId;
    [EMClient.sharedClient.contactManager declineFriendRequestFromUser:username
                                                            completion:^(NSString *aUsername, EMError *aError)
    {
        if (!aError) {
            [weakSelf onSuccess:nil callbackId:callId userInfo:nil];
        }else {
            [weakSelf onError:callId error:aError];
        }
    }];
}


- (void)deleteContact:(NSDictionary *)param callbackId:(NSString *)callbackId {
    NSString *username = param[@"username"];
    if (!username) {
        EMError *aError = [[EMError alloc] initWithDescription:@"username is invalid " code: EMErrorMessageIncludeIllegalContent];
        [self onError:callbackId error:aError];
        return;
    }
    __weak EMContactManagerWrapper * weakSelf = self;
    __block NSString *callId = callbackId;
    BOOL keepConversation = [param[@"keepConversation"] boolValue];
    [EMClient.sharedClient.contactManager deleteContact:username
                                   isDeleteConversation:keepConversation
                                          completion:^(NSString *aUsername, EMError *aError)
    {
        if (!aError) {
            [weakSelf onSuccess:nil callbackId:callId userInfo:nil];
        }else {
            [weakSelf onError:callId error:aError];
        }
    }];
}

- (id)getAllContactsFromDB:(NSDictionary *)param callbackId:(NSString *)callbackId {
    
    NSArray<NSString*>* aList = [EMClient.sharedClient.contactManager getContacts];
    if (aList == nil) {
        return nil;
    }
    return aList;
}


- (void)getAllContactsFromServer:(NSDictionary *)param callbackId:(NSString *)callbackId {
    __block NSString *callId = callbackId;
    __weak EMContactManagerWrapper * weakSelf = self;
    [EMClient.sharedClient.contactManager getContactsFromServerWithCompletion:^(NSArray *aList, EMError *aError) {
        if (!aError) {
            [weakSelf onSuccess:@"List<String>" callbackId:callId userInfo:[Transfrom NSStringFromJsonObject:aList]];
        }else {
            [weakSelf onError:callId error:aError];
        }
    }];
}

- (void)getBlockListFromServer:(NSDictionary *)param callbackId:(NSString *)callbackId {
    __weak EMContactManagerWrapper * weakSelf = self;
    __block NSString *callId = callbackId;
    [EMClient.sharedClient.contactManager getBlackListFromServerWithCompletion:^(NSArray *aList, EMError *aError) {
        if (!aError) {
            [weakSelf onSuccess:@"List<String>" callbackId:callId userInfo:[Transfrom NSStringFromJsonObject:aList]];
        }else {
            [weakSelf onError:callId error:aError];
        }
    }];
}


- (void)removeUserFromBlockList:(NSDictionary *)param callbackId:(NSString *)callbackId {
    if (!param) {
        EMError *aError = [[EMError alloc] initWithDescription:@"Param error" code: EMErrorMessageIncludeIllegalContent];
        [self onError:callbackId error:aError];
        return;
    }
    __block NSString *callId =callbackId;
    __weak EMContactManagerWrapper * weakSelf = self;
    NSString *username = param[@"username"];
    [EMClient.sharedClient.contactManager removeUserFromBlackList:username
                                                  completion:^(NSString *aUsername, EMError *aError)
    {
        if (!aError) {
            [weakSelf onSuccess:nil callbackId:callId userInfo:nil];
        }else {
            [weakSelf onError:callId error:aError];
        }
    }];
}



- (void)getSelfIdsOnOtherPlatform:(NSDictionary *)param callbackId:(NSString *)callbackId {
    __weak EMContactManagerWrapper * weakSelf = self;
    __block NSString *callId =callbackId;
    [EMClient.sharedClient.contactManager getSelfIdsOnOtherPlatformWithCompletion:^(NSArray *aList, EMError *aError) {
        if (!aError) {
            [weakSelf onSuccess:@"List<String>" callbackId:callId userInfo:aList];
        }else {
            [weakSelf onError:callId error:aError];
        }
    }];
}

@end
